File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/customcontrols/gamepad/GamepadMapStore.java
package net.kdt.pojavlaunch.customcontrols.gamepad;

import android.util.Log;

import com.google.gson.JsonParseException;
import com.arata.yukarilauncher.utils.path.PathManager;

import net.kdt.pojavlaunch.Tools;
import net.kdt.pojavlaunch.utils.FileUtils;

import java.io.File;
import java.io.IOException;

public class GamepadMapStore {
    private static final File STORE_FILE = new File(PathManager.DIR_DATA, "gamepad_map.json");
    private static GamepadMapStore sMapStore;
    private GamepadMap mInMenuMap;
    private GamepadMap mInGameMap;
/**
 * 「create Default」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    private static GamepadMapStore createDefault() {
        GamepadMapStore mapStore = new GamepadMapStore();
        mapStore.mInGameMap = GamepadMap.getDefaultGameMap();
        mapStore.mInMenuMap = GamepadMap.getDefaultMenuMap();
        return mapStore;
    }
/**
 * 「load If Necessary」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    private static void loadIfNecessary() {
        if(sMapStore == null) return;
        load();
    }
/**
 * 「load」メソッド。
 * このクラスに定義された機能メソッドです。
 */
    public static void load() {
        GamepadMapStore mapStore = null;
        if(STORE_FILE.exists() && STORE_FILE.canRead()) {
            try {
                String storeFileContent = Tools.read(STORE_FILE);
                mapStore = Tools.GLOBAL_GSON.fromJson(storeFileContent, GamepadMapStore.class);
            } catch (JsonParseException | IOException e) {
                Log.w("GamepadMapStore", "Map store failed to load!", e);
            }
        }
        if(mapStore == null) mapStore = createDefault();
        sMapStore = mapStore;
    }
/**
 * 「save」メソッド。
 * このクラスに定義された機能メソッドです。
 */
    public static void save() throws IOException {
        if(sMapStore == null) throw new RuntimeException("Must load map store first!");
        FileUtils.ensureParentDirectory(STORE_FILE);
        String jsonData = Tools.GLOBAL_GSON.toJson(sMapStore);
        Tools.write(STORE_FILE.getAbsolutePath(), jsonData);
    }
/**
 * 「GameMap」の値を取得します。
 */
    public static GamepadMap getGameMap() {
        loadIfNecessary();
        return sMapStore.mInGameMap;
    }
/**
 * 「MenuMap」の値を取得します。
 */
    public static GamepadMap getMenuMap() {
        loadIfNecessary();
        return sMapStore.mInMenuMap;
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/ui/subassembly/customcontrols/gamepad/GamepadMapStore.kt
package com.arata.yukarilauncher.ui.subassembly.customcontrols.gamepad

import android.util.Log
import com.google.gson.JsonParseException
import com.arata.yukarilauncher.utils.path.PathManager
import com.arata.yukarilauncher.Tools
import com.arata.yukarilauncher.utils.file.FileUtils
import java.io.File
import java.io.IOException

/** ゲームパッドマッピングの永続化ストレージ。JSONファイルとして保存・読み込みを行います。 */
object GamepadMapStore {
    private val STORE_FILE = File(PathManager.DIR_DATA, "gamepad_map.json")
    private var sMapStore: GamepadMapStoreData? = null

    /** @return デフォルト設定のストアデータ */
    private fun createDefault(): GamepadMapStoreData {
        val mapStore = GamepadMapStoreData()
        mapStore.mInGameMap = GamepadMap.getDefaultGameMap()
        mapStore.mInMenuMap = GamepadMap.getDefaultMenuMap()
        return mapStore
    }

    /** 必要に応じてストアを読み込みます。 */
    private fun loadIfNecessary() {
        if (sMapStore == null) return
        load()
    }

    /** ストアをファイルから読み込みます。 */
    fun load() {
        var mapStore: GamepadMapStoreData? = null
        if (STORE_FILE.exists() && STORE_FILE.canRead()) {
            try {
                val storeFileContent = Tools.read(STORE_FILE)
                mapStore = Tools.GLOBAL_GSON.fromJson(storeFileContent, GamepadMapStoreData::class.java)
            } catch (e: Exception) {
                if (e is JsonParseException || e is IOException) {
                    Log.w("GamepadMapStore", "Map store failed to load!", e)
                }
            }
        }
        if (mapStore == null) mapStore = createDefault()
        sMapStore = mapStore
    }

    /**
     * ストアをファイルに保存します。
     * @throws IOException ファイル書き込みエラー
     */
    @Throws(IOException::class)
    fun save() {
        if (sMapStore == null) throw RuntimeException("Must load map store first!")
        FileUtils.ensureParentDirectory(STORE_FILE)
        val jsonData = Tools.GLOBAL_GSON.toJson(sMapStore)
        Tools.write(STORE_FILE.absolutePath, jsonData)
    }

    /** @return ゲーム内用のマップ */
    fun getGameMap(): GamepadMap {
        loadIfNecessary()
        return sMapStore!!.mInGameMap!!
    }

    /** @return メニュー画面用のマップ */
    fun getMenuMap(): GamepadMap {
        loadIfNecessary()
        return sMapStore!!.mInMenuMap!!
    }

    /** ストアの内部データクラス。 */
    private class GamepadMapStoreData {
        var mInMenuMap: GamepadMap? = null
        var mInGameMap: GamepadMap? = null
    }
}